src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs

```diff
@@ -86,19 +86,30 @@ public async Task<string> DownloadPackageAsync(PackageId packageId,
             bool includePreview = false,
             bool? includeUnlisted = null,
             DirectoryPath? downloadFolder = null,
-            PackageSourceMapping packageSourceMapping = null)
+            PackageSourceMapping packageSourceMapping = null,
+            bool isTool = false)
         {
             CancellationToken cancellationToken = CancellationToken.None;
 
             (var source, var resolvedPackageVersion) = await GetPackageSourceAndVersion(packageId, packageVersion,
                 packageSourceLocation, includePreview, includeUnlisted ?? packageVersion is not null, packageSourceMapping).ConfigureAwait(false);
 
-            FindPackageByIdResource resource = null;
             SourceRepository repository = GetSourceRepository(source);
 
-            resource = await repository.GetResourceAsync<FindPackageByIdResource>(cancellationToken)
-                .ConfigureAwait(false);
+            if (isTool && await repository.GetResourceAsync<ServiceIndexResourceV3>().ConfigureAwait(false) is ServiceIndexResourceV3 serviceIndex)
+            {
+                var uri = serviceIndex.GetServiceEntries("SearchQueryService/3.5.0")[0].Uri;
+                var queryUri = uri + $"?q={packageId}&packageType=dotnettool";
+                using HttpClient client = new(new HttpClientHandler() { CheckCertificateRevocationList = true });
+                using HttpResponseMessage response = await client.GetAsync(queryUri).ConfigureAwait(false);
+                if (response.Content.Headers.ContentLength == 139)
+                {
+                    throw new ToolPackageException(string.Format(LocalizableStrings.NotATool, packageId));
+                }
+            }
```

Comment:
There's a strongly-typed version of the search API you can use here instead:

```suggestion
                        if (isTool && await repository.GetResourceAsync<PackageSearchResourceV3>(cancellationToken).ConfigureAwait(false) is var searchResource)
            {
                var results = await searchResource.SearchAsync(packageId.ToString(), new SearchFilter(includePrerelease: includePreview, filter: SearchFilterType.IsLatestVersion)
                {
                    PackageTypes = [NuGet.Packaging.Core.PackageType.DotnetTool.Name]
                }, skip: 0, take: 10, log: _verboseLogger, cancellationToken: cancellationToken).ConfigureAwait(false);
                if (results.Count() == 0)
                {
                    throw new ToolPackageException(string.Format(LocalizableStrings.NotATool, packageId));
                }
            }
```

A few other things to check out:
* what the error experience should be if `isTool` is true but there is no `SearchQueryService` - should we not do this and instead read the PackageMetadata and/or Registration resources first to do the check, since that should contain the `packageTypes`? cc @nkolev92 for tips/tricks to get the PackageTypes for a package correctly
* I kind of worry about NuGetPackageDownloader being used in so many different ways with different requirements. I almost want to see a `ToolPackageDownloader` that wraps the generic NuGetPackageDownloader and specializes it to make sure that tool-specific configurations/validations are done without impacting the other use cases. Same for WorkloadPackageDownloader and/or TemplatePackageDownloader, etc.

Reply:
For what it's worth, there is a ToolPackageDownloader; it's just that rather than extending/wrapping NuGetPackageDownloader, it's its own fully separate class that calls NuGetPackageDownloader when it needs to do certain things like DownloadPackageAsync. I think of it as NuGetPackageDownloader doing just that and ToolPackageDownloader providing tool-specific methods that often ultimately boil down to NuGetPackageDownloader calls. For workloads, WorkloadManifestUpdater fills a fairly similar purpose, though it goes a bit beyond workload-specific package downloading. Not sure for templates.

Reply:
fwiw, I think the NuGet API doesn't explicitly support package types right now: https://github.com/NuGet/NuGet.Client/pull/5991

Reply:
Oh awesome! Thanks for the link.

Reply:
Ok...also, `repository.GetResourceAsync<PackageSearchResourceV3>(cancellationToken).ConfigureAwait(false)` is consistently returning false, regardless of whether something is a tool or not. I verified that `https://api.nuget.org/v3/index.json` is the source it's looking at (which should have packages like dotnetsay), but it isn't working as hoped. Any idea why?